ao-tools/ao-usbtrng/ao-usbtrng.1
.\"
.\" Copyright © 2009 Keith Packard <keithp@keithp.com>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.TH AO-USBTRNG 1 "ao-usbtrng" ""
.SH NAME
ao-usbtrng \- dump random numbers from USBtrng
.SH SYNOPSIS
.B "ao-usbtrng"
[\-T \fItty-device\fP]
[\--tty \fItty-device\fP]
[\-D \fIaltos-device\fP]
[\--device \fIaltos-device\fP]
\fIkbytes\fP
.SH DESCRIPTION
.I ao-usbtrng
dumps random numbers from a USBtrng device
.SH OPTIONS
.TP
\-T tty-device | --tty tty-device
This selects which tty device the debugger uses to communicate with
the target device. The special name 'BITBANG' directs ao-dbg to use
the cp2103 connection, otherwise this should be a usb serial port
connected to a suitable cc1111 debug node.
.TP
\-D AltOS-device | --device AltOS-device
Search for a connected device. This requires an argument of one of the
following forms:
.IP
usbtrng:2
.br
usbtrng
.br
2
.IP
Leaving out the product name will cause the tool to select a suitable
product, leaving out the serial number will cause the tool to match
one of the available devices.
.SH USAGE
.I ao-usbtrng
opens the target device and reads the specified number of kbytes of
random data.
.SH AUTHOR
Keith Packard


ChangeLog
commit 26f61380ce6b4df80fa0b5a8a242cef79d5ae339
Author: Bdale Garbee <bdale@gag.com>
Date:   Sat Feb 7 22:23:38 2015 -0700

    document addition of TeleDongle to the set of ARM targets to save binaries for

commit 4231d68bae69d9a7d1f52205002db452cd5f986d
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 7 20:42:15 2015 -0800

    Bump java library versions
    
    Lots of minor API/ABI changes; make sure people can install both 1.5
    and 1.6 at the same time.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit c72d83ccd207b5300a90f2e84de6c5c96642478c
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 7 20:36:10 2015 -0800

    altosuilib: Don't offer to graph some GPS details that TM doesn't log
    
    When using a TM eeprom file, various minor GPS details are logged
    (course, ground speed, climb rate, etc). Make sure these aren't
    offered up for graphing.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 426bacbbd6e0573b143e7c48d71db977e53181fc
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 7 20:22:19 2015 -0800

    altosuilib: Get the Eeprom download progress bar working again
    
    The eeprom download code wasn't computing the start of each state
    transition correctly, so the progress bar was snapping to the end of
    the chunk for each state.
    
    Invalid state values would snap the bar to the right side.
    
    Landed state wasn't ever seen, so the bar would not ever fill.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit aea188e4227993c508c7045d02f2bc87abd4e428
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 7 18:10:34 2015 -0800

    doc: update 1.6 release notes with recent bug fixes
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 0dd733efdbb66535889aafecbba45f49266ed886
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 7 17:49:39 2015 -0800

    altosuilib: Set the icon and title of the telemetry scanning dialog
    
    Just need to call 'super' to get the window configured correctly.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 9d5a6914982a29d22103e0202735c9e01ee9e831
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 7 17:35:15 2015 -0800

    altoslib: Clear saved TelemetryReader state instead of resetting in ScanUI
    
    Instead of attempting to mash the saved state from AltosScanUI, just
    throw it away in AltosTelemetryReader.reset. Much cleaner, and more reliable.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit cbc9ecd960d0dd27a2275b4550c30c5c88ccf5bb
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 7 17:09:40 2015 -0800

    altosuilib: Reset telemetry state after processing it
    
    This avoids re-using stale state after switching telemetry reception
    parameters around, which otherwise generates spurious entries for the
    wrong frequencies in the scan results list.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit c0ce45100dd205f59cd84465138c74d9b31df0bb
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 7 17:08:03 2015 -0800

    altoslib: Drop telemetry packets processed while monitoring is disabled
    
    A lag between the thread queuing telemetry packets and one pulling
    them out can result in stale telemetry data being returned to the
    reader. Fix this by dropping telemetry read while monitoring is disabled.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 8a2be4d36a3b116d82529805430c5fb665688267
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 7 15:09:57 2015 -0800

    altosuilib: Wait for product data while scanning
    
    Now that Altos only sends config information once every 5 seconds, we
    want to keep waiting for that if we're receiving packets successfully.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit e70f3dca01b15b75b0b8795eb71bd12817af4800
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 7 14:43:11 2015 -0800

    altoslib: Mark listener as 'not running' on EOF.
    
    This adds a 'running' member to the AltosListenerState class, and when
    the replay reader reaches EOF, marks the listener as no longer
    running.
    
    AltosUI and TeleGPS now display 'done' in the 'Age' field when this
    occurs, to let the user know that the replay is over.
    
    Also make sure that the display timers are stopped when this happens,
    or when the window is closed.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 08f07d0af123e1c307bc4c0c973da07fae8246b1
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 7 14:40:17 2015 -0800

    altosui: Remove the dregs of AltosDroid load-old-telem code
    
    AltosDroid used to scan the old .telem file to return to the previous
    flight state on restart. Now it just loads the old state object
    instead, a vast improvement in performance.
    
    To do that, there were some changes in the altoslib code
    required. This patch just removes those, fixing replay bugs in TeleGPS
    along the way.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 0d08c427188f7b748d9ea7651143bc843e9c2691
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 7 13:09:18 2015 -0800

    altos: ao_packet_getchar needs to return 'int' so AO_READ_AGAIN works
    
    With ao_packet_getchar returning 'char', AO_READ_AGAIN would get
    converted to 0xff instead of -1 on return, making ao_packet_echo
    either produce a bogus character on output, breaking command
    responses, or wedge and never exit when the link was dropped.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 4a5c64bef6952f25cadebc3e1f9909a5f513e015
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Feb 7 02:37:36 2015 -0800

    Update version 1.6 release notes with more bug fixes
    
    A few bugs found during release testing are notable enough to be
    included in the release notes.
    

ao-tools/ao-usbtrng/ao-usbtrng.c
/*
 * Copyright © 2014 Keith Packard <keithp@keithp.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

#include <err.h>
#include <fcntl.h>
#include <gelf.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <sysexits.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include <stdbool.h>
#include "ao-elf.h"
#include "ccdbg.h"
#include "cc-usb.h"
#include "cc.h"
#include "ao-verbose.h"

static const struct option options[] = {
	{ .name = "tty", .has_arg = 1, .val = 'T' },
	{ .name = "device", .has_arg = 1, .val = 'D' },
	{ .name = "raw", .has_arg = 0, .val = 'r' },
	{ .name = "verbose", .has_arg = 1, .val = 'v' },
	{ 0, 0, 0, 0},
};

static void usage(char *program)
{
	fprintf(stderr, "usage: %s [--verbose=<verbose>] [--device=<device>] [-tty=<tty>] <kbytes>\n", program);
	exit(1);
}

void
done(struct cc_usb *cc, int code)
{
	cc_usb_close(cc);
	exit (code);
}

int
main (int argc, char **argv)
{
	char			*device = NULL;
	char			*filename;
	Elf			*e;
	unsigned int		s;
	int			i;
	int			c;
	int			tries;
	struct cc_usb		*cc = NULL;
	char			*tty = NULL;
	int			success;
	int			verbose = 0;
	int			ret = 0;
	int			expected_size;
	int			kbytes;
	uint8_t			bits[1024];

	while ((c = getopt_long(argc, argv, "rT:D:c:s:v:", options, NULL)) != -1) {
		switch (c) {
		case 'T':
			tty = optarg;
			break;
		case 'D':
			device = optarg;
			break;
		case 'v':
			verbose++;
			break;
		default:
			usage(argv[0]);
			break;
		}
	}

	if (!argv[optind])
		usage(argv[0]);

	kbytes = atoi(argv[optind]);
	if (kbytes < 1)
		kbytes = 1;

	ao_verbose = verbose;

	if (verbose > 1)
		ccdbg_add_debug(CC_DEBUG_BITBANG);

	if (!tty)
		tty = cc_usbdevs_find_by_arg(device, "usbtrng");
	if (!tty)
		tty = getenv("ALTOS_TTY");
	if (!tty)
		tty="/dev/ttyACM0";

	cc = cc_usb_open(tty);

	if (!cc)
		exit(1);

	cc_usb_printf(cc, "f %d\n", kbytes);

	while (kbytes--) {
		int	i;
		for (i = 0; i < 1024; i++)
			bits[i] = cc_usb_getchar(cc);
		write(1, bits, 1024);
	}

	done(cc, ret);
}
